src/tools/perf/perftest.c

```diff
@@ -1585,31 +1588,33 @@ static ucs_status_t run_test_recurs(struct perftest_context *ctx,
         return UCS_ERR_IO_ERROR;
     }
 
-    status = clone_params(&params, parent_params);
-    if (status != UCS_OK) {
-        goto out;
-    }
-
     line_num = 0;
-    while ((status = read_batch_file(batch_file, ctx->batch_files[depth],
-                                     &line_num, &params,
-                                     &ctx->test_names[depth])) == UCS_OK) {
-        run_test_recurs(ctx, &params, depth + 1);
-        free(params.super.msg_size_list);
-        free(ctx->test_names[depth]);
-        ctx->test_names[depth] = NULL;
-
+    for (;;) {
         status = clone_params(&params, parent_params);
         if (status != UCS_OK) {
             goto out;
         }
+
+        status = read_batch_file(batch_file, ctx->batch_files[depth],
+                                 &line_num, &params,
+                                 &ctx->test_names[depth]);
+        if (status != UCS_OK) {
+            break;
+        }
+
+        run_test_recurs(ctx, &params, depth + 1);
+        free(params.super.msg_size_list);
+        params.super.msg_size_list = NULL;
+        free(ctx->test_names[depth]);
+        ctx->test_names[depth] = NULL;
     }
 
     if (status == UCS_ERR_NO_ELEM) {
         status = UCS_OK;
     }
 
     free(params.super.msg_size_list);
+    params.super.msg_size_list = NULL;
```

Comment:
maybe we don't need the free() in line 1616 now? and also this? since it's don'e in lines 1605-1606?

Reply:
yes, we could avoid this `free()`, but need to reorganize code a bit
done